master/internal/db/postgres_tasks.go

```diff
@@ -119,18 +119,19 @@ WHERE task_id = $1
 	return nil
 }
 
-// KillGenericTask persists the termination of a task of type GENERIC.
-func (db *PgDB) KillGenericTask(tID model.TaskID, endTime time.Time) error {
-	err := completeTask(db.sql, tID, endTime)
-	if err != nil {
-		return err
-	}
+// TerminateGenericTask persists the termination of a task of type GENERIC.
+func (db *PgDB) TerminateGenericTask(tID model.TaskID, endTime time.Time) error {
 	if _, err := db.sql.Exec(`
-UPDATE tasks
-SET task_state = $2
-WHERE task_id = $1
-	`, tID, model.TaskStateCanceled); err != nil {
-		return errors.Wrap(err, "killing task")
+	UPDATE tasks
+	SET task_state = (
+	CASE WHEN task_state=$2 THEN $3::task_state
+	ELSE $4::task_state END),
+	end_time = (CASE WHEN task_state=$5 THEN $6::timestamp ELSE NULL END)
```

Comment:
1. is there a risk this is wiping out `end_time` field? perhaps `(CASE WHEN task_state=$5 THEN $6::timestamp ELSE end_time END)`
2. I think we'd want to record the end time for paused tasks as well, e.g. when task goes from StoppingPaused -> Paused, why don't we record its end time as well?
3. it seems like for commands we set end_time in the OnExit callback regardless of the state: https://github.com/determined-ai/determined/blob/main/master/internal/command/command.go#L233 ; so, why do we need a state conditional here? which state transitions is it supposed to catch?

Reply:
If we want to have endtime for paused we can remove the `end_time = (CASE WHEN task_state=$5 THEN $6::timestamp ELSE NULL END)` and we wont risk wiping out end_time. I initially thought we only want end_time for tasks that are completed.

For conditions on the OnExit, when a task completes we want to mark that task_state as `COMPLETED`. When we pause or kill a task the allocation will terminate however we want that task state to be `PAUSED` or `CANCELED`. Both scenarios return the same "terminated" response from the allocation so we need to do an additional check on line 126 of this function.

Reply:
we currently set end_time for tasks which are killed. I think we should set it for pause action as well, because the end results after kill and pause is the same: the container is killed and it's not running any more. effectively it's the timestamp of the last "exit" event on it...